WooCommerce/WooCommerceUITests/Screens/Login/PrologueScreen.swift

```diff
@@ -1,20 +1,27 @@
-import Foundation
+import ScreenObject
 import XCTest
 
 private struct ElementStringIDs {
     static let continueButton = "Prologue Continue Button"
     static let siteAddressButton = "Prologue Self Hosted Button"
 }
 
-final class PrologueScreen: BaseScreen {
-    private let continueButton: XCUIElement
-    private let siteAddressButton: XCUIElement
+final class PrologueScreen: ScreenObject {
+    private let continueButtonGetter: (XCUIApplication) -> XCUIElement = {
+        $0.buttons[ElementStringIDs.continueButton]
+    }
```

Comment:
In this implementation, I experimented with computing elements at runtime every time, instead of only once.

Now that I've sat on it for a bit, I wonder if that's a good idea or not.

The reason I went for it in the first place was to ensure we always had up to date elements, but I do wonder if that's really necessary given `XCUIElement` is a `NSObject` subclass, that is, an object and therefore assigned by reference. 🤔

Reply:
IMHO fetching everything every time is preferable because some screens may not _always_ contain every UI element (consider a table view that might not contain an actual tableview if there's no data, but instead have a nice picture in the center)

Reply:
I don't remember where I wrote about it, but I worry that by running a query on every access, we'll make the tests considerably slower, because of how slow querying `XCUIApplication` is.

But I do agree with your point, which is exactly why I went for computing in the first place.

🤔 I wonder if there's a good way to measure this.

---

Regarding `continueButton` vs `continueButtonGetter`. Here, it's called `-Getter` because otherwise it would name-clash with the actual `continueButton` `XCUIElement` below.

Looking at the code, I also noticed that the `-Getter`s are unused, which is a mistake on my side. A leftover from my experimented mentioned below, on which I then took a step back (as you can see in the [source](https://github.com/Automattic/ScreenObject/blob/df8ee4983e674b583b546bb0640ec83428be2465/Sources/ScreenObject/ScreenObject.swift) for screen object.

I do agree on this with you, too. If we were to go down the "always compute" path, I would have something like this

```swift
// PrologueScreen.swift
var continueButton: XCUIElement { expectedElement }
var otherButton: XCUIElement { app.buttons["other button"] }
// where `app` is the instance property inherited from the superclass, as discussed

// ScreenObject.swift
var expectedElement: XCUIElement { expectedElementGetter(app) }
```

Reply:
Good catch. I didn't clean up properly before enabling auto-merge 🤦‍♂️ 

👉 https://github.com/woocommerce/woocommerce-ios/pull/4770